rand intn in loop

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
OS: windows, Microsoft Windows 10 Home v19044 64-bit
Processor: 12 cpus, 64bit, little endian, Intel(R) Core(TM) i5-10600K CPU @ 4.10GHz
CC version: Error: exec failed (CreateProcess) with code 2: The system cannot find the file specified.
 cmd: cc --version

getwd: D:\code\v
vmodules: C:\Users\epper\.vmodules
vroot: D:\code\v
vexe: D:\code\v\v.exe
vexe mtime: 2022-04-25 12:37:37
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 88f22b4.e42dc8e

Git version: git version 2.29.2.windows.3
Git vroot status: weekly.2022.16-62-ge42dc8e2 (2 commit(s) behind V master)
.git/config present: true
thirdparty/tcc status: thirdparty-windows-amd64 0137a863


<!-- Please include the standalone minimal reproducible code.
If not, it may be closed by the administrators. -->
```
// trying to recreate: http://rosettacode.org/wiki/Average_loop_length#Go
import rand
import math {abs}

const nmax = 20

fn ana(n int) f64 {
    return 1.0 // haven't started
}
fn avg(n int) f64 {
    tests := 1e4
    mut sum := 0
    for t := 0; t < tests; t++ {
        mut v := []bool{len: nmax, init: false}
        for x := 0; !v[x]; x = rand.intn(n) or {0} { //THIS LINE..Adding () doesn't chane anything
            v[x] = true
            sum++
        }
    }
    return f64(sum)/tests
}
fn main() {
    println(' N   average   analytical   (error)')
    println('=== ========= ============ =========')
    for n in 1 .. nmax+1 {
        a := avg(n)
        b := ana(n)
        println('${n:3} ${a:9.4f} ${b:12.4f} (${(abs(a-b)/b*100):6.2f}%)')
    }
}
```
**What did you do?**
`./v run ../rosetta/average_loop.v`

**What did you expect to see?**
Something like:
```
 N    average    analytical    (error)
===  =========  ============  =========
  1     1.0000        1.0000  (  0.00%)
  2     1.5007        1.5000  (  0.05%)
...
```

 
**What did you see instead?**
```
C:/Users/epper/AppData/Local/Temp/v_0/average_loop.7237628358776190551.tmp.c:473: warning: WINVER redefined
C:/Users/epper/AppData/Local/Temp/v_0/average_loop.7237628358776190551.tmp.c:7307: warning: implicit declaration of function 'tcc_backtrace'
C:/Users/epper/AppData/Local/Temp/v_0/average_loop.7237628358776190551.tmp.c:7926: warning: assignment from incompatible pointer type
C:/Users/epper/AppData/Local/Temp/v_0/average_loop.7237628358776190551.tmp.c:7926: warning: assignment makes pointer from integer without a cast
C:/Users/epper/AppData/Local/Temp/v_0/average_loop.7237628358776190551.tmp.c:7926: warning: cast between pointer and integer of different size
C:/Users/epper/AppData/Local/Temp/v_0/average_loop.7237628358776190551.tmp.c:13532: warning: assignment from incompatible pointer type
C:/Users/epper/AppData/Local/Temp/v_0/average_loop.7237628358776190551.tmp.c:13539: warning: assignment from incompatible pointer type
C:/Users/epper/AppData/Local/Temp/v_0/average_loop.7237628358776190551.tmp.c:13546: warning: assignment from incompatible pointer type
C:/Users/epper/AppData/Local/Temp/v_0/average_loop.7237628358776190551.tmp.c:13554: warning: assignment from incompatible pointer type
C:/Users/epper/AppData/Local/Temp/v_0/average_loop.7237628358776190551.tmp.c:13582: warning: assignment from incompatible pointer type
C:/Users/epper/AppData/Local/Temp/v_0/average_loop.7237628358776190551.tmp.c:13582: warning: assignment from incompatible pointer type
C:/Users/epper/AppData/Local/Temp/v_0/average_loop.7237628358776190551.tmp.c:13602: warning: assignment from incompatible pointer type
C:/Users/epper/AppData/Local/Temp/v_0/average_loop.7237628358776190551.tmp.c:13604: warning: assignment from incompatible pointer type
C:/Users/epper/AppData/Local/Temp/v_0/average_loop.7237628358776190551.tmp.c:13604: warning: assignment from incompatible pointer type
C:/Users/epper/AppData/Local/Temp/v_0/average_loop.7237628358776190551.tmp.c:13624: warning: assignment from incompatible pointer type
C:/Users/epper/AppData/Local/Temp/v_0/average_loop.7237628358776190551.tmp.c:16694: error: ')' expected (got "_t1")
builder error:
==================
C error. This should never happen.

```

--- a/vlib/v/checker/for.v
+++ b/vlib/v/checker/for.v
@@ -17,7 +17,7 @@ fn (mut c Checker) for_c_stmt(node ast.ForCStmt) {
 			for right in node.inc.right {
 				if right is ast.CallExpr {
 					if right.or_block.stmts.len > 0 {
-						c.error('disallow using optional call in conditions of for_c_stmt',
+						c.error('optionals are not allowed in `for` statement condition',
 							right.pos)
 					}
 				}

--- a/vlib/v/checker/for.v
+++ b/vlib/v/checker/for.v
@@ -13,6 +13,16 @@ fn (mut c Checker) for_c_stmt(node ast.ForCStmt) {
 	}
 	c.expr(node.cond)
 	if node.has_inc {
+		if node.inc is ast.AssignStmt {
+			for right in node.inc.right {
+				if right is ast.CallExpr {
+					if right.or_block.stmts.len > 0 {
+						c.error('disallow using optional call in conditions of for_c_stmt',
+							right.pos)
+					}
+				}
+			}
+		}
 		c.stmt(node.inc)
 	}
 	c.check_loop_label(node.label, node.pos)

--- a/vlib/v/checker/for.v
+++ b/vlib/v/checker/for.v
@@ -17,7 +17,7 @@ fn (mut c Checker) for_c_stmt(node ast.ForCStmt) {
 			for right in node.inc.right {
 				if right is ast.CallExpr {
 					if right.or_block.stmts.len > 0 {
-						c.error('optionals are not allowed in `for` statement condition (yet)',
+						c.error('optionals are not allowed in `for statement increment` (yet)',
 							right.pos)
 					}
 				}

--- a/vlib/v/checker/for.v
+++ b/vlib/v/checker/for.v
@@ -17,7 +17,7 @@ fn (mut c Checker) for_c_stmt(node ast.ForCStmt) {
 			for right in node.inc.right {
 				if right is ast.CallExpr {
 					if right.or_block.stmts.len > 0 {
-						c.error('optionals are not allowed in `for` statement condition',
+						c.error('optionals are not allowed in `for` statement condition, it is not implemented yet',
 							right.pos)
 					}
 				}

--- a/vlib/v/checker/for.v
+++ b/vlib/v/checker/for.v
@@ -17,7 +17,7 @@ fn (mut c Checker) for_c_stmt(node ast.ForCStmt) {
 			for right in node.inc.right {
 				if right is ast.CallExpr {
 					if right.or_block.stmts.len > 0 {
-						c.error('optionals are not allowed in `for` statement condition, it is not implemented yet',
+						c.error('optionals are not allowed in `for` statement condition (yet)',
 							right.pos)
 					}
 				}

--- a/vlib/v/checker/tests/for_c_stmt_with_optional_call.out
+++ b/vlib/v/checker/tests/for_c_stmt_with_optional_call.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/for_c_stmt_with_optional_call.vv:15:31: error: disallow using optional call in conditions of for_c_stmt
+   13 |     for t := 0; t < tests; t++ {
+   14 |         mut v := []bool{len: nmax, init: false}
+   15 |         for x := 0; !v[x]; x = rand.intn(n) or { 0 } {
+      |                                     ~~~~~~~
+   16 |             v[x] = true
+   17 |             sum++

--- a/vlib/v/checker/tests/for_c_stmt_with_optional_call.out
+++ b/vlib/v/checker/tests/for_c_stmt_with_optional_call.out
@@ -1,4 +1,4 @@
-vlib/v/checker/tests/for_c_stmt_with_optional_call.vv:15:31: error: optionals are not allowed in `for` statement condition, it is not implemented yet
+vlib/v/checker/tests/for_c_stmt_with_optional_call.vv:15:31: error: optionals are not allowed in `for` statement condition (yet)
    13 |     for t := 0; t < tests; t++ {
    14 |         mut v := []bool{len: nmax, init: false}
    15 |         for x := 0; !v[x]; x = rand.intn(n) or { 0 } {

--- a/vlib/v/checker/tests/for_c_stmt_with_optional_call.out
+++ b/vlib/v/checker/tests/for_c_stmt_with_optional_call.out
@@ -1,4 +1,4 @@
-vlib/v/checker/tests/for_c_stmt_with_optional_call.vv:15:31: error: optionals are not allowed in `for` statement condition (yet)
+vlib/v/checker/tests/for_c_stmt_with_optional_call.vv:15:31: error: optionals are not allowed in `for statement increment` (yet)
    13 |     for t := 0; t < tests; t++ {
    14 |         mut v := []bool{len: nmax, init: false}
    15 |         for x := 0; !v[x]; x = rand.intn(n) or { 0 } {

--- a/vlib/v/checker/tests/for_c_stmt_with_optional_call.out
+++ b/vlib/v/checker/tests/for_c_stmt_with_optional_call.out
@@ -1,4 +1,4 @@
-vlib/v/checker/tests/for_c_stmt_with_optional_call.vv:15:31: error: disallow using optional call in conditions of for_c_stmt
+vlib/v/checker/tests/for_c_stmt_with_optional_call.vv:15:31: error: optionals are not allowed in `for` statement condition
    13 |     for t := 0; t < tests; t++ {
    14 |         mut v := []bool{len: nmax, init: false}
    15 |         for x := 0; !v[x]; x = rand.intn(n) or { 0 } {

--- a/vlib/v/checker/tests/for_c_stmt_with_optional_call.out
+++ b/vlib/v/checker/tests/for_c_stmt_with_optional_call.out
@@ -1,4 +1,4 @@
-vlib/v/checker/tests/for_c_stmt_with_optional_call.vv:15:31: error: optionals are not allowed in `for` statement condition
+vlib/v/checker/tests/for_c_stmt_with_optional_call.vv:15:31: error: optionals are not allowed in `for` statement condition, it is not implemented yet
    13 |     for t := 0; t < tests; t++ {
    14 |         mut v := []bool{len: nmax, init: false}
    15 |         for x := 0; !v[x]; x = rand.intn(n) or { 0 } {

--- a/vlib/v/checker/tests/for_c_stmt_with_optional_call.vv
+++ b/vlib/v/checker/tests/for_c_stmt_with_optional_call.vv
@@ -1,0 +1,31 @@
+import rand
+import math { abs }
+
+const nmax = 20
+
+fn ana(n int) f64 {
+	return 1.0 // haven't started
+}
+
+fn avg(n int) f64 {
+	tests := 1e4
+	mut sum := 0
+	for t := 0; t < tests; t++ {
+		mut v := []bool{len: nmax, init: false}
+		for x := 0; !v[x]; x = rand.intn(n) or { 0 } {
+			v[x] = true
+			sum++
+		}
+	}
+	return f64(sum) / tests
+}
+
+fn main() {
+	println(' N   average   analytical   (error)')
+	println('=== ========= ============ =========')
+	for n in 1 .. nmax + 1 {
+		a := avg(n)
+		b := ana(n)
+		println('${n:3} ${a:9.4f} ${b:12.4f} (${(abs(a - b) / b * 100):6.2f}%)')
+	}
+}

